File: src/main/java/com/sivalabs/bookmarker/domain/service/UserService.java
package com.sivalabs.bookmarker.domain.service;

import com.sivalabs.bookmarker.domain.entity.Role;
import com.sivalabs.bookmarker.domain.entity.User;
import com.sivalabs.bookmarker.domain.exception.BookmarkerException;
import com.sivalabs.bookmarker.domain.exception.ResourceNotFoundException;
import com.sivalabs.bookmarker.domain.model.ChangePasswordRequest;
import com.sivalabs.bookmarker.domain.model.UserDTO;
import com.sivalabs.bookmarker.domain.repository.RoleRepository;
import com.sivalabs.bookmarker.domain.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.Optional;

import static com.sivalabs.bookmarker.domain.utils.Constants.ROLE_USER;

@Service
@Transactional
@Slf4j
@RequiredArgsConstructor
public class UserService {
    private final UserRepository userRepository;
    private final RoleRepository roleRepository;
    private final PasswordEncoder passwordEncoder;

    @Transactional(readOnly = true)
    public Optional<UserDTO> getUserById(Long id) {
        return userRepository.findById(id).map(UserDTO::fromEntity);
    }

    @Transactional(readOnly = true)
    public Optional<User> getUserByEmail(String email) {
        return userRepository.findByEmail(email);
    }

    @Transactional(readOnly = true)
    public boolean isUserExistsByEmail(String email) {
        return userRepository.existsByEmail(email);
    }

    public UserDTO createUser(UserDTO user) {
        if (userRepository.existsByEmail(user.getEmail())) {
            throw new BookmarkerException("Email "+user.getEmail()+" is already in use");
        }
        user.setPassword(passwordEncoder.encode(user.getPassword()));
        User userEntity = user.toEntity();
        Optional<Role> roleUser = roleRepository.findByName(ROLE_USER);
        userEntity.setRoles(Collections.singleton(roleUser.orElse(null)));
        return UserDTO.fromEntity(userRepository.save(userEntity));
    }

    public UserDTO updateUser(UserDTO user) {
        Optional<User> byId = userRepository.findById(user.getId());
        if(byId.isEmpty()) {
            throw new ResourceNotFoundException("User with id "+user.getId()+" not found");
        }
        User userEntity = user.toEntity();
        userEntity.setPassword(byId.get().getPassword());
        userEntity.setRoles(byId.get().getRoles());
        return UserDTO.fromEntity(userRepository.save(userEntity));

    }

    public void deleteUser(Long userId) {
        Optional<User> byId = userRepository.findById(userId);
        byId.ifPresent(userRepository::delete);
    }

    public void changePassword(String email, ChangePasswordRequest changePasswordRequest) {
        Optional<User> userByEmail = this.getUserByEmail(email);
        if(userByEmail.isEmpty()) {
            throw new ResourceNotFoundException("User with email "+email+" not found");
        }
        User user = userByEmail.get();
        if (passwordEncoder.matches(changePasswordRequest.getOldPassword(), user.getPassword())) {
            user.setPassword(passwordEncoder.encode(changePasswordRequest.getNewPassword()));
            userRepository.save(user);
        } else {
            throw new BookmarkerException("Current password doesn't match");
        }
    }
}

File: backend/src/main/kotlin/com/sivalabs/bookmarker/domain/service/UserService.kt
package com.sivalabs.bookmarker.domain.service

import com.sivalabs.bookmarker.domain.exception.BookmarkerException
import com.sivalabs.bookmarker.domain.exception.UserNotFoundException
import com.sivalabs.bookmarker.domain.entity.User
import com.sivalabs.bookmarker.domain.repository.RoleRepository
import com.sivalabs.bookmarker.domain.repository.UserRepository
import com.sivalabs.bookmarker.domain.model.ChangePasswordRequest
import com.sivalabs.bookmarker.domain.model.UserDTO
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.Optional

@Service
@Transactional
class UserService(
    private val userRepository: UserRepository,
    private val roleRepository: RoleRepository,
    private val passwordEncoder: PasswordEncoder
) {

    @Transactional(readOnly = true)
    fun getUserById(id: Long): Optional<UserDTO> {
        return userRepository.findById(id).map { UserDTO.fromEntity(it) }
    }

    @Transactional(readOnly = true)
    fun getUserByEmail(email: String): Optional<User> {
        return userRepository.findByEmail(email)
    }

    fun createUser(user: UserDTO): UserDTO {
        if (userRepository.existsByEmail(user.email)) {
            throw BookmarkerException("Email ${user.email} is already in use")
        }
        user.password = passwordEncoder.encode(user.password)
        val userEntity = user.toEntity()
        roleRepository.findByName("ROLE_USER").map { userEntity.roles = mutableListOf(it) }
        return UserDTO.fromEntity(userRepository.save(userEntity))
    }

    fun updateUser(user: UserDTO): UserDTO {
        return userRepository.findById(user.id).map {
            val userEntity = user.toEntity()
            userEntity.password = it.password
            userEntity.roles = it.roles
            UserDTO.fromEntity(userRepository.save(userEntity))
        }.orElseThrow { UserNotFoundException("User with id ${user.id} not found") }
    }

    fun deleteUser(userId: Long) {
        userRepository.findById(userId).map { userRepository.delete(it) }
    }

    fun changePassword(email: String, changePasswordRequest: ChangePasswordRequest) {
        this.getUserByEmail(email).map { user ->
            if (passwordEncoder.matches(changePasswordRequest.oldPassword, user.password)) {
                user.password = passwordEncoder.encode(changePasswordRequest.newPassword)
                userRepository.save(user)
            } else {
                throw BookmarkerException("Current password doesn't match")
            }
        }.orElseThrow { UserNotFoundException("User with email $email not found") }
    }
}
